### DISCUSSION TITLE
library - general

### DISCUSSION THREAD

**Discussion by ivanhofer** [pipeli18ne/RFC#3] [Feedback]
Comments about the proposed solution to the library problem

**Comment by samuelstroschein** [comment]
I am not confident whether the proposed "one library to rule them via plugins" has high adoption feasibility. 

The benefit of _slightly_ reducing the learning curve by minimizing the requirement to use different i18n libraries does not seem high enough to drive larger adoption. Furthermore, I question the motivation itself. Most i18n library consumers have just two interfaces: 

```tsx
// 1. a translate function 
t("hello")

// 2. a component for markup translations
<Trans>...</Trans>
```

What happens in the background is, in most cases, not a concern of the consumer e.g. if a developer is in a React codebases or Svelte codebase makes no difference. It does make a difference _when something does not work_. In the latter case, letting developers write a tiny bit of boilerplate (looking at the SvelteKit example) is likely a good thing. If something does not work, and the implementation is "magically" hidden, the cost of not writing a few lines of boilerplate outweigh the benefits of being able to quickly fix or adapt a solution for a specific requirement. 

The SvelteKit example from the RFC reveals another reason why such a package might be a dead end. I know that SvelteKit wants to build its own i18n solution to deliver the best possible experience. This decision comes down to "How easy can we [Svelte] build an i18n solution to deliver a good experience for SvelteKit users?". The answer is: trivial if other frameworks do not need to be considered. _Even more trivial_ if something like a `link-layer` exists to build upon.  

While the idea of one i18n library to rule them all is noble, the complexity of unifying different frameworks and edge-cases for little upside and also competing against "choose the library specifically made for react|svelte|next-js|solid" who will provide a better experience for the particular framework, seems not worth to pursue.

**Reply by ivanhofer** [reply]
Thanks for your feedback @samuelstroschein!

I agree that the `library` approach is the hardest to convince people to use if they now existing solutions. While the `link-layer` is certainly the core aspect, the `plugin` approach would be a great addition that could highly improve the developer experience when working with i18n.
 - It is a key aspect to make [`build-runtime optimizations`](https://github.com/pipeli18ne/RFC/blob/main/02-library.md#build-time-optimizations) possible.
 - Supporting both [`key-based` and `template-based` translations simultaneously](https://github.com/pipeli18ne/RFC/blob/main/02-library.md#key-based-translations-vs-template-based-translations) is also something I haven't seen in a i18n library before.
 - It makes also [use-cases like this](https://github.com/ivanhofer/typesafe-i18n/discussions/448) and [others]() possible.

The `link-layer` will make it possible for other libraries to support such use-cases, but then there is the need for a wrapper for each of those features for each library. And then we are again staying in front of the same problem: not all libraries will support all features.

To your other concern: you can opt-out any time from the "magic" those plugins provide ([_I tried to make this more clear in the RFC_](https://github.com/pipeli18ne/RFC/blob/main/02-library.md#opt-out-of-code-injection)). It is a good starting point that will probably cover >90% of use cases. If it does not fit your needs, you can manually write the code you need. If the plugin "sees" you have written manually code, it will not auto-inject anything. Like in `Svelte` you can use stores by manually subscribing to them, or you use the `$`-sign syntax as a convenient feature.

**Reply by samuelstroschein** [reply]
In that case, rephrasing the benefit of such a library in the RFC seems advisable. From "slightly reducing learning curve" to "enabling further optimization". 

My overall sentiment of mine remains. Further optimizations and additional features are nice, but they likely won't drive adoption. A basic i18n library with no build-runtime that loads a JSON resource is good enough for the majority of apps. If no server-side rendering is involved, build-runtime optimization benefits further diminish.

**Reply by ivanhofer** [reply]
The `optimization` is not really the key aspect of this chapter. It actually is "make i18n more accessible" by eliminating boilerplate code, reducing learning curves and making the hurdle to add i18n and working with it as small as possible.

I have altered the "motivation" section slightly.

**Reply by osdiab** [reply]
What's the goal? Personally not all libraries need to be everything to all users. If you're writing a client-side only app then yeah, I agree that simplifying your build pipeline is a good thing, but at least for people like me who do have more requirements than that—and there are many of us—I would be keen for a unified approach to i18n that provides a great DX and is efficient for the underlying platform.

I see this as a more complex solution that indeed may be overkill for a really simple use case, and that's not a bad thing per-se.

**Reply by ivanhofer** [reply]
The goal is to create a library that is easy to use, making it more accessible to developers who start with i18n. At the same time it should allow more advanced use cases if someone needs it. If you are using `React` in your frontend, the same i18n experience should be carried over to your `Node.js` backend and your `Solid,js` side project. You probably don't want to spend some time figuring out different libraries and the tooling around them.
In my opinion a "classical library" is not the best fit. Therefore I have thought of using `plugins` that can help improve DX. If someone wishes, he can also use the plain `library`, but needs to figure out everything on their own.

It has also to do with my mindset of "doing things once". Why should every project need to figure out how to change their code if they switch from `Next.js` version 12 to 13. Let a group of people come up for a good solution and release a new version of that plugin. Then every project can benefit from it by simply upgrading to the new plugin.

**Comment by osdiab** [comment]
This may not be a mass-market appealing thing, but thinking about the key- vs template-based translations philosophy, a thought:

In the world of Graphql, with the newer versions of `graphql-codegen`, you write your graphql queries using the `gql` template literal directly where you use the code. Then the codegen reads all your source files and generates a library that hooks up all those queries flexibly to the client library you choose, with a typesafe interface. It's neat.

Benefits of the template-based approach include that the actual text is right where your code is, and that as you showed with that builder pattern, you could potentially do so in a very type-safe way (no fat-fingered invalid translation strings).

Mashing these two ideas together it seems like it would be feasible to take the same approach as something like graphql-codegen, to scan and convert these invocations of a template (to allow for nice DX) to a more standard i18n format (to allow for ease of translation by nontechnical translators, and seamlessly interop with solutions like inlang, lokalize, crowdin etc). I've not thought about this problem as much as you have but seems like it could be a nice DX and that there's prior art.

**Reply by ivanhofer** [reply]
In my opinion `key-based` translations don't scale when you have a really big project. As we all know, naming things is hard and coming up with a good key for each string is not that easy. So why not let tooling do that for you. Just write your strings into your components and let the [`adapters`](https://github.com/pipeli18ne/RFC/blob/main/01-link-layer.md#string-extraction) do the rest.

If someone uses a `template-based` approach, I think types are not really necessary since you write your code directly in your template and don't use the abstraction like in the case of a `key-based` translation function.

But the approach of "developers should just write actual code and the tooling does the rest" matches with the `graphql-codegen` methodology.
